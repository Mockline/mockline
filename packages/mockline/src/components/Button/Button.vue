<script setup lang="ts">
import { twMerge } from 'tailwind-merge'

export type ButtonProps = {
  /** Button label */
  label: string;
  /** Button class */
  class?: string;
  /** Block button */
  block?: boolean;
}

const props = defineProps<ButtonProps>()

const defaultClass = 'bg-blue-600 px-2 py-1 rounded text-white'

const block = props.block ? 'w-full' : 'w-fit'

const buttonClass = twMerge(defaultClass, props.class, block)
</script>

<template>
  <div>
    <div :class="buttonClass">
      {{ props.label }}
    </div>
  </div>
</template>

<style scoped>

</style>
