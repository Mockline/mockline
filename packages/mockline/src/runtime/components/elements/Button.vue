<script lang="ts">
import { tv, type VariantProps } from 'tailwind-variants'
import type { AppConfig } from '@nuxt/schema'
import { computed } from 'vue'
import { twMerge } from 'tailwind-merge'
import _appConfig from '#build/app.config'
import theme from '#build/mockline/button'
</script>

<script setup lang="ts">

const appConfig = _appConfig as AppConfig & { mockline: { button: Partial<typeof theme> } }

const button = tv({ extend: tv(theme), ...(appConfig.mockline?.button || {}) })

type ButtonVariants = VariantProps<typeof button>

export type ButtonProps = {
  label?: string
  color?: ButtonVariants['color']
  variant?: ButtonVariants['variant']
  size?: ButtonVariants['size']
  square?: boolean
  block?: boolean
  class?: any
  ui?: Partial<typeof button.slots>
  icon?: string
  loading?: boolean
  leading?: boolean
  trailing?: boolean
}

export type ButtonSlots = {
  leading(): any
  default(): any
  trailing(): any
}

const props = withDefaults(defineProps<ButtonProps>(), {
  leading: true,
  trailing: false,
})

const slots = defineSlots<ButtonSlots>()

const ui = computed(() => tv({ extend: button, slots: props.ui })({
  color: props.color,
  variant: props.variant,
  size: props.size,
  loading: props.loading,
  block: props.block,
  square: props.square || (!slots.default && !props.label),
}))
</script>

<template>
  <component
    :is="'button'"
    :class="twMerge(ui.base({ class: props.class }))"
  >
    <slot name="leading">
      <MIcon v-if="leading && loading" name="lucide:loader" :class="ui.leadingIcon(props)" />
      <MIcon v-if="leading && icon" :name="icon" :class="ui.leadingIcon(props)" />
    </slot>

    <span v-if="label || !!slots.default" :class="ui.label()">
      <slot>
        {{ label }}
      </slot>
    </span>

    <slot name="trailing">
      <MIcon v-if="trailing && loading" name="lucide:loader" :class="ui.trailingIcon(props)" />
      <MIcon v-if="trailing && icon" :name="icon" :class="ui.trailingIcon(props)" />
    </slot>
  </component>
</template>
