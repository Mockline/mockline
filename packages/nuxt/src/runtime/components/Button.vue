<script setup lang="ts">
import { computed } from 'vue'
import { type ButtonProps, type ButtonSlots } from '@mockline/themes'
import { useComponent } from '../utils/useComponent'
import { useComponentIcons, type UseComponentIconsProps } from '#mockline/composables/useComponentIcons'
import appConfig from '#build/app.config'

const props = defineProps<ButtonProps & UseComponentIconsProps>()
const slots = defineSlots<ButtonSlots>()

const { isLeading, isTrailing, leadingIconName, trailingIconName } = useComponentIcons(props)

const componentProps = computed(() => {
  return {
    ...props,
    square: props.square || (!slots.default && !props.label),
    leading: isLeading.value,
    trailing: isTrailing.value,
    transitions: appConfig.mockline.transitions
  }
})

const { getClasses } = useComponent('button', componentProps)
</script>

<template>
  <component
    :is="'button'"
    :class="getClasses('base', props.class)"
  >
    <slot name="leading">
      <MIcon
        v-if="isLeading && leadingIconName"
        :name="leadingIconName"
        :class="getClasses('leadingIcon', iconClass)"
      />
    </slot>

    <span v-if="label || !!slots.default" :class="getClasses('label', labelClass)">
      <slot>
        {{ label }}
      </slot>
    </span>

    <slot name="trailing">
      <MIcon
        v-if="isTrailing && trailingIconName"
        :name="trailingIconName"
        :class="getClasses('trailingIcon', iconClass)"
      />
    </slot>
  </component>
</template>
