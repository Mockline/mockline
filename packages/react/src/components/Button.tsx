import {FC, ReactNode} from "react";
import {Icon} from './Icon.tsx';
import {useComponent} from '../utils/useComponent.tsx';
import {useComponentIcons, UseComponentIconsProps} from '../composables/useComponentIcons.tsx';

type ButtonProps = {
    square?: boolean;
    label?: string;
    className?: string;
    iconClass?: string;
    labelClass?: string;
    children?: ReactNode;
    leading?: ReactNode;
    trailing?: ReactNode;
} & UseComponentIconsProps;

export const Button: FC<ButtonProps> = (props: ButtonProps) => {
    const {isLeading, isTrailing, leadingIconName, trailingIconName} = useComponentIcons(props)

    const componentProps = {
        ...props,
        square: props.square || (!props.children && !props.label),
        leading: isLeading,
        trailing: isTrailing,
    };

    const {getClasses} = useComponent("button", componentProps);

    return (
        <button className={getClasses("base", props.className)}>
            {(typeof props.leading != 'boolean' && props.leading) ||
                (isLeading && leadingIconName && (
                    <Icon name={leadingIconName} className={getClasses("leadingIcon", props.iconClass)}/>
                ))
            }

            {props.children || (props.label && (
                <span className={getClasses("label", props.labelClass)}>
                  {props.children || props.label}
                </span>
            ))}

            {(typeof props.trailing != 'boolean' && props.trailing) ||
                (isTrailing && trailingIconName && (
                    <Icon name={trailingIconName} className={getClasses("trailingIcon", props.iconClass)}/>
                ))}
        </button>
    );
};
