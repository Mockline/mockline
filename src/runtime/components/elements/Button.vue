<script setup lang="ts">
import { twMerge } from 'tailwind-merge'

export type ButtonProps = {
  /** Button label */
  label: string;
  /** Button class */
  class?: string;
  /** Block button */
  block?: boolean;
  /** Button type */
  type?: 'button' | 'submit' | 'reset';
  /** Size */
  size?: 'sm' | 'md' | 'lg' | 'xl';
}

const props = defineProps<ButtonProps>()

const defaultClass = 'bg-blue-600 px-2 py-1 rounded text-white'

const block = computed(() => props.block ? 'w-full' : 'w-fit')

const size = computed(() => {
  switch (props.size) {
    case 'sm':
      return 'text-sm'
    case 'md':
      return 'text-md'
    case 'lg':
      return 'text-lg'
    case 'xl':
      return 'text-xl'
    default:
      return 'text-md'
  }
})

const buttonClass = computed(() => twMerge(defaultClass, props.class, block.value, size.value))
</script>

<template>
  <component :is="'button'" :class="buttonClass" :type>
    {{ props.label }}
  </component>
</template>

<style scoped>

</style>
