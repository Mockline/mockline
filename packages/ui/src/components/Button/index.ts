import Button from './Button.vue'
import type { ButtonProps } from './Button.vue'

export { Button, type ButtonProps }
