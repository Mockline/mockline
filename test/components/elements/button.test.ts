import { describe, it, expect } from 'vitest'
import Button, { type ButtonProps } from '../../../src/runtime/components/elements/Button.vue'
import ComponentRender from '../component-render'

describe('Button', () => {
  it('renders correctly', async () => {
    const html = await ComponentRender('default', {}, Button)
    expect(html).toMatchSnapshot()
  })
  /*it.each([
    ['with label', { props: { label: 'Button' } }],
    ['with size', { props: { size: 'sm' } }],
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: ButtonProps, slots?: any }): Promise<void> => {
    const html = await ComponentRender(nameOrHtml, options, Button)
    expect(html).toMatchSnapshot()
  })*/
})
