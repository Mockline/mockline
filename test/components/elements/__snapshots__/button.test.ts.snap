// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders <Button icon="i-heroicons-pencil-square" size="sm" color="primary" square variant="solid" /> correctly 1`] = `"<button class="bg-blue-600 px-2 py-1 rounded text-white w-fit text-sm" icon="i-heroicons-pencil-square" color="primary" square="" variant="solid"></button>"`;

exports[`Button > renders basic case correctly 1`] = `"<button class="bg-blue-600 px-2 py-1 rounded text-white w-fit text-md"></button>"`;

exports[`Button > renders black solid correctly 1`] = `"<button class="bg-blue-600 px-2 py-1 rounded text-white w-fit text-md" color="black" variant="solid"></button>"`;

exports[`Button > renders rounded full correctly 1`] = `"<button class="bg-blue-600 px-2 py-1 rounded text-white w-fit text-md" ui="[object Object]"></button>"`;
